### ISSUE TITLE
"[Err] XRUN" and audio cutting out

### ISSUE THREAD

**Issue by PureTryOut** [badaix/snapcast#496]
I'm running snapclient on a Raspberry Pi 3B, but it reports lots of `[Err] XRUN` and the music cuts out constantly (I literally only hear milliseconds of music every 1 or so second), making it unusable. Normally I listen with Snapclient on my desktop or laptop where it works fine, and I believe it has also worked on the RPi in the past.

Any clue what is happening here? I want to use this setup for a party within a month or so but this makes it a bit difficult.

**Comment by badaix** [comment]
Have you tried to use a different sample rat, i.e. 48000 instead of 44100?
Some samplerates are natively supported by the DSP while others are not, I think this will also be logged as a warning.

**Comment by PureTryOut** [comment]
Actually it already used 48000 by default (just running `snapserver`). However after trying out 44100, the issue persisted.

**Comment by badaix** [comment]
Is there anything suspicious in the client logs? Can you test with aplay if alsa is generally working on your system?

**Comment by PureTryOut** [comment]
That RPi system currently already runs a networked PulseAudio system (receiving client streams and playing it on it's own speakers) and that works fine, so alsa in generally works.

Besides the error I mentioned, no nothing suspicious is in the client logs.

**Comment by badaix** [comment]
Maybe it helps to switch the soundcard by passing an index returned by `-l` to the `-s` option.

**Comment by PureTryOut** [comment]
There is only one soundcard that works on my setup and that is the PulseAudio output, all the others indicate "Device or resource busy" (I'm guessing PulseAudio is blocking it).

**Comment by PureTryOut** [comment]
I'd like to this setup coming saturday for a party, is there anyone else I can try?

I do have a fallback device to use where it works fine, but I rather use the RPi.

**Comment by badaix** [comment]
What os are you using?
I'm testing with raspian and snapos, both are running fine on RPi 3B

**Comment by PureTryOut** [comment]
Raspbian 9, RPi3B.

**Comment by badaix** [comment]
I have a RPi 3b+ (Hardware: BCM2835, Revision: a020d3)
```
pi@raspberrypi:~ $ cat /proc/cpuinfo 
processor	: 0
model name	: ARMv7 Processor rev 4 (v7l)
BogoMIPS	: 89.60
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4

processor	: 1
model name	: ARMv7 Processor rev 4 (v7l)
BogoMIPS	: 89.60
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4

processor	: 2
model name	: ARMv7 Processor rev 4 (v7l)
BogoMIPS	: 89.60
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4

processor	: 3
model name	: ARMv7 Processor rev 4 (v7l)
BogoMIPS	: 89.60
Features	: half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xd03
CPU revision	: 4

Hardware	: BCM2835
Revision	: a020d3
Serial		: 000000008810b4fb
```

OS
```
pi@raspberrypi:~ $ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 9 (stretch)"
NAME="Raspbian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
VERSION_CODENAME=stretch
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

Kernel
```
pi@raspberrypi:~ $ uname -a
Linux raspberrypi 4.19.66-v7+ #1253 SMP Thu Aug 15 11:49:46 BST 2019 armv7l GNU/Linux
```



Soundcards: (Snapclient uses the one with `default` in it's name as default DSP)
```
pi@raspberrypi:~ $ snapclient -l
0: null
Discard all samples (playback) or generate zero samples (capture)

1: default:CARD=ALSA
bcm2835 ALSA, bcm2835 ALSA
Default Audio Device

2: sysdefault:CARD=ALSA
bcm2835 ALSA, bcm2835 ALSA
Default Audio Device

3: dmix:CARD=ALSA,DEV=0
bcm2835 ALSA, bcm2835 ALSA
Direct sample mixing device

4: dmix:CARD=ALSA,DEV=1
bcm2835 ALSA, bcm2835 IEC958/HDMI
Direct sample mixing device

5: dmix:CARD=ALSA,DEV=2
bcm2835 ALSA, bcm2835 IEC958/HDMI1
Direct sample mixing device

6: dsnoop:CARD=ALSA,DEV=0
bcm2835 ALSA, bcm2835 ALSA
Direct sample snooping device

7: dsnoop:CARD=ALSA,DEV=1
bcm2835 ALSA, bcm2835 IEC958/HDMI
Direct sample snooping device

8: dsnoop:CARD=ALSA,DEV=2
bcm2835 ALSA, bcm2835 IEC958/HDMI1
Direct sample snooping device

9: hw:CARD=ALSA,DEV=0
bcm2835 ALSA, bcm2835 ALSA
Direct hardware device without any conversions

10: hw:CARD=ALSA,DEV=1
bcm2835 ALSA, bcm2835 IEC958/HDMI
Direct hardware device without any conversions

11: hw:CARD=ALSA,DEV=2
bcm2835 ALSA, bcm2835 IEC958/HDMI1
Direct hardware device without any conversions

12: plughw:CARD=ALSA,DEV=0
bcm2835 ALSA, bcm2835 ALSA
Hardware device with all software conversions

13: plughw:CARD=ALSA,DEV=1
bcm2835 ALSA, bcm2835 IEC958/HDMI
Hardware device with all software conversions

14: plughw:CARD=ALSA,DEV=2
bcm2835 ALSA, bcm2835 IEC958/HDMI1
Hardware device with all software conversions
```


Initial debug log
```
pi@raspberrypi:~ $ snapclient --debug
2019-11-19 19-38-00.693 [Info] (browse_callback) (Browser) NEW: service 'Snapcast' of type '_snapcast._tcp' in domain 'local'
2019-11-19 19-38-00.694 [Info] (browse_callback) (Browser) CACHE_EXHAUSTED
2019-11-19 19-38-00.695 [Info] (resolve_callback) Service 'Snapcast' of type '_snapcast._tcp' in domain 'local':
2019-11-19 19-38-00.695 [Info] (resolve_callback) 	raspberrypi.local:1704 (192.168.0.3)
2019-11-19 19-38-00.695 [Debug] (resolve_callback) 	TXT=
...
2019-11-19 19-38-00.695 [Debug] (resolve_callback) 	cached: 1
2019-11-19 19-38-00.697 [Info] (main) Found server 192.168.0.3:1704
2019-11-19 19-38-00.697 [Info] (main) Latency: 0
2019-11-19 19-38-00.697 [Debug] (start) Connecting
2019-11-19 19-38-00.698 [Notice] (start) Connected to 192.168.0.3
2019-11-19 19-38-00.698 [Info] (getMacAddress) My MAC: "b8:27:eb:10:b4:fb", socket: 6
2019-11-19 19-38-00.711 [Debug] (onMessageReceived) time sync onMessageReceived
2019-11-19 19-38-00.957 [Info] (onMessageReceived) ServerSettings - buffer: 1000, latency: 0, volume: 51, muted: 0
metadata:{"STREAM":"default"}
2019-11-19 19-38-00.957 [Info] (onMessageReceived) Codec: flac
2019-11-19 19-38-00.957 [Notice] (state) sampleformat: 48000:16:2
2019-11-19 19-38-00.958 [Debug] (setVolume_exp) setVolume exp: 0.51 => 0.248437
2019-11-19 19-38-00.962 [Debug] (initAlsa) PCM name: default:CARD=ALSA
2019-11-19 19-38-00.963 [Debug] (initAlsa) PCM state: PREPARED
2019-11-19 19-38-00.963 [Debug] (initAlsa) channels: 2
2019-11-19 19-38-00.963 [Debug] (initAlsa) rate: 48000 bps
2019-11-19 19-38-00.963 [Info] (initAlsa) frames: 1440
2019-11-19 19-38-00.963 [Debug] (initAlsa) period time: 30000
```

"Normal runtime" log (no XRUNs)
```
2019-11-19 19-40-17 [Info] Chunk: 0	0	1	1	500	116
2019-11-19 19-40-18 [Info] Chunk: 0	0	0	1	500	116
2019-11-19 19-40-19 [Info] Chunk: -1	-1	0	1	500	116
2019-11-19 19-40-20 [Info] Chunk: -1	-1	-1	0	500	116
2019-11-19 19-40-21 [Info] Chunk: 1	0	-1	0	500	116
2019-11-19 19-40-22 [Info] Chunk: 1	1	0	0	500	116
2019-11-19 19-40-23 [Info] Chunk: 1	1	1	0	500	116
2019-11-19 19-40-24 [Info] Chunk: 0	0	1	0	500	116
2019-11-19 19-40-25 [Info] Chunk: -1	-1	0	0	500	116
2019-11-19 19-40-26 [Info] Chunk: -1	-1	0	0	500	116
2019-11-19 19-40-27 [Info] Chunk: -1	-1	-1	0	500	116
2019-11-19 19-40-28 [Info] Chunk: 1	1	-1	0	500	116
2019-11-19 19-40-29 [Info] Chunk: 2	1	0	0	500	116
2019-11-19 19-40-30 [Info] Chunk: 1	2	1	0	500	116
2019-11-19 19-40-31 [Info] Chunk: 0	0	1	0	500	116
2019-11-19 19-40-32 [Info] Chunk: -1	-1	0	0	500	116
2019-11-19 19-40-33 [Info] Chunk: 0	-1	0	0	500	116
2019-11-19 19-40-34 [Info] Chunk: -1	-1	-1	0	500	116
2019-11-19 19-40-35 [Info] Chunk: 0	0	-1	0	500	116
2019-11-19 19-40-36 [Info] Chunk: 1	1	0	0	500	116
2019-11-19 19-40-37 [Info] Chunk: 1	1	1	0	500	116
2019-11-19 19-40-38 [Info] Chunk: 0	0	1	0	500	116
2019-11-19 19-40-39 [Info] Chunk: -1	-1	0	0	500	116
2019-11-19 19-40-40 [Info] Chunk: 0	-1	0	0	500	116
2019-11-19 19-40-41 [Info] Chunk: -1	-1	-1	0	500	116
2019-11-19 19-40-42 [Info] Chunk: 1	1	-1	0	500	116
```

**Comment by PureTryOut** [comment]
Hardware (Raspberry Pi 3B, BCM2835, revision a02082):
```
▶ cat /proc/cpuinfo 
processor       : 0
model name      : ARMv7 Processor rev 4 (v7l)
BogoMIPS        : 38.40
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

processor       : 1
model name      : ARMv7 Processor rev 4 (v7l)
BogoMIPS        : 38.40
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

processor       : 2
model name      : ARMv7 Processor rev 4 (v7l)
BogoMIPS        : 38.40
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

processor       : 3
model name      : ARMv7 Processor rev 4 (v7l)
BogoMIPS        : 38.40
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt vfpd32 lpae evtstrm crc32 
CPU implementer : 0x41
CPU architecture: 7
CPU variant     : 0x0
CPU part        : 0xd03
CPU revision    : 4

Hardware        : BCM2835
Revision        : a02082
Serial          : 00000000485a2968
```

OS:
```
▶ cat /etc/os-release 
PRETTY_NAME="Raspbian GNU/Linux 9 (stretch)"
NAME="Raspbian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
VERSION_CODENAME=stretch
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

Kernel:
```
▶ uname -a
Linux raspberrypi 4.14.98-v7+ #1200 SMP Tue Feb 12 20:27:48 GMT 2019 armv7l GNU/Linux
```

Soundcards (default is used, others get `Device or resource busy`):
```
▶ snapclient -l
0: default
Playback/recording through the PulseAudio sound server

1: null
Discard all samples (playback) or generate zero samples (capture)

2: sysdefault:CARD=Device
USB Sound Device, USB Audio
Default Audio Device

3: front:CARD=Device,DEV=0
USB Sound Device, USB Audio
Front speakers

4: surround21:CARD=Device,DEV=0
USB Sound Device, USB Audio
2.1 Surround output to Front and Subwoofer speakers

5: surround40:CARD=Device,DEV=0
USB Sound Device, USB Audio
4.0 Surround output to Front and Rear speakers

6: surround41:CARD=Device,DEV=0
USB Sound Device, USB Audio
4.1 Surround output to Front, Rear and Subwoofer speakers

7: surround50:CARD=Device,DEV=0
USB Sound Device, USB Audio
5.0 Surround output to Front, Center and Rear speakers

8: surround51:CARD=Device,DEV=0
USB Sound Device, USB Audio
5.1 Surround output to Front, Center, Rear and Subwoofer speakers

9: surround71:CARD=Device,DEV=0
USB Sound Device, USB Audio
7.1 Surround output to Front, Center, Side, Rear and Woofer speakers

10: iec958:CARD=Device,DEV=0
USB Sound Device, USB Audio
IEC958 (S/PDIF) Digital Audio Output

11: dmix:CARD=Device,DEV=0
USB Sound Device, USB Audio
Direct sample mixing device

12: dsnoop:CARD=Device,DEV=0
USB Sound Device, USB Audio
Direct sample snooping device

13: hw:CARD=Device,DEV=0
USB Sound Device, USB Audio
Direct hardware device without any conversions

14: plughw:CARD=Device,DEV=0
USB Sound Device, USB Audio
Hardware device with all software conversions
```

Initial debug log:
```
▶ snapclient -h <address> --debug
2019-11-19 23-11-25.371 [Info] (main) Latency: 0
2019-11-19 23-11-25.374 [Debug] (start) Connecting
2019-11-19 23-11-25.383 [Notice] (start) Connected to 85.149.153.230
2019-11-19 23-11-25.384 [Info] (getMacAddress) My MAC: "b8:27:eb:5a:29:68", socket: 6
2019-11-19 23-11-25.668 [Info] (onMessageReceived) ServerSettings - buffer: 1000, latency: 0, volume: 100, muted: 0
metadata:{"STREAM":"Mopidy"}
2019-11-19 23-11-25.669 [Info] (onMessageReceived) Codec: flac
2019-11-19 23-11-25.669 [Notice] (state) sampleformat: 48000:16:2
2019-11-19 23-11-25.669 [Debug] (setVolume_exp) setVolume exp: 1 => 1
2019-11-19 23-11-25.675 [Debug] (initAlsa) PCM name: default
2019-11-19 23-11-25.675 [Debug] (initAlsa) PCM state: PREPARED
2019-11-19 23-11-25.675 [Debug] (initAlsa) channels: 2
2019-11-19 23-11-25.675 [Debug] (initAlsa) rate: 48000 bps
2019-11-19 23-11-25.675 [Info] (initAlsa) frames: 1440
2019-11-19 23-11-25.675 [Debug] (initAlsa) period time: 30000
```

The strange is that I recall this setup working fine in the past (and still does on other devices), I'm not sure what changed.

**Comment by badaix** [comment]
Maybe you can tweak the period time, which is set to 30000: `[Debug] (initAlsa) period time: 30000`. The period time is some interval when to check for new data, once the audio buffer is full (i.e. should be smaller than the buffer), see [here](https://stackoverflow.com/questions/24040672/the-meaning-of-period-in-alsa).
The period time is hard coded in `client/player/alsa_player.cpp`, line 25.
I'm still wondering if the problem is reproducible with `aplay`, there the period time is also configurable with `--period-time`. 
How is your CPU usage during playback?

**Comment by badaix** [comment]
Btw: I'm using raspian lite, which seems not to have a Pulseaudio server installed.

**Comment by PureTryOut** [comment]
I just tested a .wav file with aplay, works perfectly fine with default settings and `--period-time 30000` as well.

CPU usage while running snapclient and playing something over it is reported by `htop` to be ~35%.

PulseAudio wasn't installed by default either here, but I installed it manually so I can use it both for network usage (e.g. playing laptop and desktop audio over the network speakers) and as a Bluetooth audio receiver for my phone.
